### PR TITLE
feat: Detect and log orphaned or errored gradle project scans by comparing detected files vs scanned.

### PR DESCRIPTION
**Pull request by lili2311**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team
### Depends on
- https://github.com/snyk/snyk-cli-interface/pull/38
- https://github.com/snyk/snyk-gradle-plugin/pull/142
#### What does this PR do?
Detect and log orphaned or errored gradle project scans by comparing detected files vs scanned.
This is useful to find any gradle files that looked like they are part of a multi-project build but were in fact not included or had an error being scanned. Since only the root level file is detected these errors can otherwise fo unnoticed.

#### Where should the reviewer start?
https://github.com/snyk/snyk/pull/1355/files#diff-91534f3e6e36be30e5b2863969dc6417R82

#### How should this be manually tested?
Drop a gradle file in a folder in a multi-build project, this file is not included as a sub-project and will be detected but not scanned since the gradle build will not return it.

#### Screenshots
Screen Shot 2020-08-21 at 11 43 52

### CODE REVIEW

File: src/lib/plugins/get-deps-from-plugin.ts [hunk 1]
```diff
@@ -105,3 +118,43 @@ export async function getDepsFromPlugin(
...
+
+function warnSomeGradleManifestsNotScanned(
+  scannedProjects: ScannedProject[],
+  allFilesFound: string[],
+  root: string,
+  optionsJson = false,
+): void {
+  const scannedGradleFiles = scannedProjects
+    .map((p) => {
+      const targetFile = p.meta?.targetFile || p.targetFile;
```
[the remark this answers is not in the document]
Comment (lili2311) [reply src/lib/plugins/get-deps-from-plugin.ts:130]: I see what you mean ty! Will keep this in mind, in this particular PR we first auto detect any `build.gradle` files and then compare the detected vs scanned to alert if some were missed for any reason.

File: src/lib/find-files.ts [hunk 2]
```diff
@@ -102,20 +123,30 @@ async function findInDirectory(
...
-        return [];
+        return { files: [], allFilesFound: [] };
       }
       return find(resolvedPath, ignore, filter, levelsDeep);
     });
+
   const found = await Promise.all(toFind);
-  return Array.prototype.concat.apply([], found);
+  return {
+    files: Array.prototype.concat.apply(
```
Comment (aviadhahami) [inline src/lib/find-files.ts:141]: using `apply` in combination w/ `map` will make this call allocate 3 arrays; imo we can simply return the mapped array:
```suggestion
    files: found.map(...)
```


> I'm not sure why this specific trick was previously used, as the `this` arg for `concat.apply` is an empty array so the "append" magic `concat.apply` can do is redundant;
 [`concat.apply` example](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply#Using_apply_to_append_an_array_to_another)

File: src/lib/plugins/get-deps-from-plugin.ts [hunk 3]
```diff
@@ -105,3 +121,40 @@ export async function getDepsFromPlugin(
...
+      return targetFile ? pathLib.resolve(root, targetFile) : null;
+    })
+    .filter(
+      (targetFile) =>
+        targetFile &&
+        (targetFile.endsWith('build.gradle') ||
+          targetFile.endsWith('build.gradle.kts')),
+    );
+  const detectedGradleFiles = allFilesFound.filter(
+    (targetFile) =>
```
Comment (aviadhahami) [inline src/lib/plugins/get-deps-from-plugin.ts:142]: we need to assert `targetFile` as well no?

Comment (aviadhahami) [reply src/lib/plugins/get-deps-from-plugin.ts:142]: Can we extract the following code?
```js
      (targetFile) =>
        targetFile &&
        (targetFile.endsWith('build.gradle') ||
          targetFile.endsWith('build.gradle.kts'))
```

something like:
```js
const gradleTargetFilesFilter = (targetFile)=> (...);
...
const scannedGradleFiles = scannedProjects.map(...).filter(gradleTargetFilesFilter)
const detectedGradleFiles = allFilesFound.filter(gradleTargetFilesFilter)

File: src/lib/find-files.ts (hunk 2 above)
Comment (lili2311) [reply src/lib/find-files.ts:141]: The result is nested arrays, `concat.apply` flattens them all into 1 array.
Screen Shot 2020-08-26 at 17 14 08

Comment (lili2311) [reply src/lib/find-files.ts:141]: I think I can use `.flat()` will try it!

Comment (lili2311) [reply src/lib/find-files.ts:141]: @aviadatsnyk .flat() is not Node 8 compatible and couldn't get it working with TS with es2019 either :/ so reverted back for now until we drop Node 8

**Review by aviadhahami — APPROVED** [review]
Lgtm :)

**Review by aviadhahami — APPROVED** [review]
LGTM :checkered_flag:

**Pull request merged**